contracts/schemes/Competition.sol

```diff
@@ -0,0 +1,349 @@
+pragma solidity 0.5.13;
+
+import "./ContributionRewardExt.sol";
+
+
+contract Competition {
+    using SafeMath for uint256;
+
+    uint256 constant public MAX_NUMBER_OF_WINNERS = 100;
+
+    event NewCompetitionProposal(
+        bytes32 indexed _proposalId,
+        uint256 _numberOfWinners,
+        uint256[] _rewardSplit,
+        uint256 _startTime,
+        uint256 _votingStartTime,
+        uint256 _endTime,
+        uint256 _numberOfVotesPerVoters,
+        address payable _contributionRewardExt //address of the contract to redeem from.
+    );
+
+    event Redeem(
+        bytes32 indexed _proposalId,
+        uint256 indexed _suggestionId,
+        uint256 _rewardPercentage
+    );
+
+    event NewSuggestion(
+        bytes32 indexed _proposalId,
+        uint256 indexed _suggestionId,
+        string indexed _descriptionHash,
+        address payable suggester
+    );
+
+    event NewVote(
+        bytes32 indexed _proposalId,
+        uint256 indexed _suggestionId,
+        address indexed _voter,
+        uint256 _reputation
+    );
+
+    event SnapshotBlock(
+        bytes32 indexed _proposalId,
+        uint256 _snapshotBlock
+    );
+
+    // A struct holding the data for a competition proposal
+    struct Proposal {
+        uint256 numberOfWinners;
+        uint256[] rewardSplit;
+        uint256 startTime;
+        uint256 votingStartTime;
+        uint256 endTime;
+        uint256 numberOfVotesPerVoters;
+        address payable contributionRewardExt;
+        uint256 snapshotBlock;
+        //contributionRewardExt
+        uint256 reputationReward;
+        uint256 ethReward;
+        uint256 nativeTokenReward;
+        uint256 externalTokenReward;
+
+        uint256[] topSuggestionsOrdered;
+        //mapping from suggestions totalVotes to the number of suggestions with the same totalVotes.
+        mapping(uint256=>uint256) tiesSuggestions;
```

Comment:
Actually, this mapping will  lso store suggestions that have a unique number of votes - so not just ties. Perhaps rename to `votesToSuggestions` or `suggestionsPerVote`.